segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriterFactory.java

```diff
@@ -114,7 +115,12 @@ public UpdateableSegmentMetadata getStreamSegmentMetadata(long streamSegmentId)
 
         @Override
         public InputStream getAppendData(long streamSegmentId, long startOffset, int length) {
-            return this.readIndex.readDirect(streamSegmentId, startOffset, length);
+            try {
+                return this.readIndex.readDirect(streamSegmentId, startOffset, length);
+            } catch (StreamSegmentNotExistsException ex) {
+                // Null is interpreted as "Segment not exists" by the SegmentAggregator.
```

Comment:
Looking at the callers and the other methods on this interface: It might be better to change the signature so that it throws this exception. SegmentAggrigator responds by failing a future or throwing a new exception in the event of null, so it might as well be able to preserve the full stack trace.

Reply:
Yes, I did consider that, but inside the Read Index I do a check for the segment's existence first, and then I look up that in that index for the data. If the segment gets deleted exactly after the check but before retrieving the data, we'll be in trouble. Handling that kind of a race properly is more complicated than aggregating this into one.

Plus, the stack trace of `StreamSegmentNotExistsException` is not terribly useful - it itself indicates the segment doesn't exist; where it originates from does not really help much.